File: java/953_VerifyingAnAlienDictionary.java
class Solution {
    HashMap<Character, Integer> orderDict;
    public boolean isAlienSorted(String[] words, String order) {
        orderDict = new HashMap<Character, Integer>();
        for (int i = 0; i < order.length(); ++i) {
            orderDict.put(order.charAt(i), i);
        }
        
        String prev = words[0];
        for (int i = 1; i < words.length; ++i) {
            if (!check(prev, words[i])) return false;
            prev = words[i];
        }
        return true;
    }
    
    public boolean check(String w1, String w2) {
        if ((w1.isEmpty() || w2.isEmpty())) {
            return w1.length() <= w2.length();
        } else if (orderDict.get(w1.charAt(0)) < orderDict.get(w2.charAt(0))) {
            return true;
        } else if (orderDict.get(w1.charAt(0)) > orderDict.get(w2.charAt(0))) {
            return false;
        } else {
            return check(w1.substring(1, w1.length()), w2.substring(1, w2.length()));
        }
    }
}

File: python/953_VerifyingAnAlienDictionary.py
class Solution:
    def isAlienSorted(self, words: List[str], order: str) -> bool:
        order_dict = {}
        for i, l in enumerate(order):
            order_dict[l] = i
        
        def check(w1, w2):
            if not (w1 and w2):
                return len(w1) <= len(w2)
            elif order_dict[w1[0]] < order_dict[w2[0]]:
                return True
            elif order_dict[w1[0]] > order_dict[w2[0]]:
                return False
            else:
                return check(w1[1:], w2[1:])
        
        prev = words[0]
        for i in range(1, len(words)):
            if not check(prev, words[i]):
                return False
            prev = words[i]
        return True
